projects/adc_periodic_reader/inc/adc_periodic_reader.h

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "adc.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt.h"
+#include "log.h"
+#include "misc.h"
+#include "soft_timer.h"
+#include "status.h"
+
+typedef enum {
+  PERIODIC_READER_ID_0 = 0,
+  PERIODIC_READER_ID_1,
+  PERIODIC_READER_ID_2,
+  PERIODIC_READER_ID_3,
+  PERIODIC_READER_ID_4,
+  NUM_PERIODIC_READER_IDS
+} PeriodicReaderId;
+
+typedef void (*AdcPeriodicReaderCallback)(uint16_t data, PeriodicReaderId id, void *context);
+
+typedef struct {
+  GpioAddress address;
+  AdcPeriodicReaderCallback callback;
+  bool activated;
+} AdcPeriodicReaderSettings;
+
+typedef struct {
+  uint16_t data;
+} AdcPeriodicReaderStorage;
+
+static AdcPeriodicReaderSettings s_settings[NUM_PERIODIC_READER_IDS];
+static AdcPeriodicReaderStorage s_storage[NUM_PERIODIC_READER_IDS];

```

Comment:
keep everything in s_storage, when they pass in `settings` you copy their stuff from `settings` to `storage`.